### ISSUE TITLE
R&D profiling

### ISSUE THREAD

**Issue by Plyb** [DRAGNLabs/MIRROR-Pipeline#203]
**Acceptance Criteria**:
- a report exists on what our options are for profiling our code's performance
- Things we want:
  - Something like a flame/icicle chart showing what functions/spans the program is spending a lot of time in
  - Works when we're doing multi-node training
    - probably just needs to run on global rank zero
  - Ideally, we can instrument our code and have it remain clean and not have a performance hit when profiling isn't running
    - decorators might be useful here?
    - But if we can't do this, the next best thing is that it is easy to add and remove the profiling.

**Comment by morrisoncheney** [comment]
Based on my research and convos with Claude torch.profiler seems to me to be the best option.
Here is a doc that Claude made to explain it.

---

# Profiling the MIRROR Pipeline

## Options at a Glance

| Tool | Best for | Setup cost |
|------|----------|------------|
| `cProfile` + `snakeviz` | Full program call-graph, finding which functions are called most | Low — built into Python |
| `line_profiler` | Line-by-line timing inside a specific function | Low — `pip install line_profiler` |
| `torch.profiler` | GPU/CPU kernel timing, memory, data loading vs. compute breakdown | Low — built into PyTorch |
| `time.perf_counter()` | Quick sanity-check timing of a specific block | Zero — built into Python |

For a training pipeline, **`torch.profiler` is almost always the right first tool** — it understands GPU operations and shows you whether time is lost in compute, data loading, or CPU↔GPU synchronization.

---

## torch.profiler

### What it tells you

- How long each PyTorch operation takes on CPU and GPU
- Whether the GPU is idle waiting on the CPU (a common bottleneck)
- Memory usage over time
- Which operations are called most and dominate total time

### Basic usage

Wrap any block of code in the context manager:

```python
import torch

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
) as prof:
    # code to profile

print(prof.key_averages().table(sort_by="cuda_time_total", row_limit=20))
```

This prints a table like:

```
---------------------------------  --------  --------  --------
Name                               CPU total  CUDA total  Calls
---------------------------------  --------  --------  --------
aten::linear                         1.2ms     8.4ms      48
aten::embedding                      0.3ms     0.1ms      12
aten::scaled_dot_product_attention   0.8ms     6.1ms      12
...
```

Sort by `cuda_time_total` to find GPU bottlenecks, or `cpu_time_total` to find CPU-bound operations.

### Viewing as a visual timeline

Export a Chrome trace after profiling:

```python
prof.export_chrome_trace("trace.json")
```

Open `chrome://tracing` in Chrome and load the file. This gives a lane-per-thread timeline showing exactly when each kernel ran and where the CPU and GPU were idle.

---

## Using torch.profiler in the MIRROR Trainer

The main training loop lives in [`trainer.py`](../../src/mirror/trainer.py) at lines 165–181. Everything important — the forward pass, backward pass, optimizer step, and dataloader iteration — happens there.

### Profiling a sample of batches

Profiling every batch of a full run is slow and produces an unmanageably large trace. Use `torch.profiler.schedule` to capture only a representative window:

```python
with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.CUDA,
    ],
    schedule=torch.profiler.schedule(wait=5, warmup=1, active=3),
    on_trace_ready=torch.profiler.tensorboard_trace_handler("./profiler_logs"),
) as prof:
    for batch_idx, batch in batch_iter:
        optimizer.zero_grad()
        train_step_output = model.training_step(batch)
        loss_value = train_step_output.loss.item()
        self.fabric.backward(train_step_output.loss)
        optimizer.step()
        if scheduler is not None:
            scheduler.step()

        global_step = epoch_idx * n_batches + batch_idx
        self.fabric.call('on_train_batch_end', ...)

        prof.step()  # must be called at the end of each batch
```

The schedule parameters:
- `wait=5` — skip the first 5 batches (let things warm up)
- `warmup=1` — profile but discard the next 1 batch (GPU caches warming)
- `active=3` — capture the next 3 batches (these are what you'll analyze)

### What to look for

Once you have results, the key questions are:

1. **Is `cuda_time_total` >> `cpu_time_total` for the forward pass?** Good — GPU is doing the work.
2. **Are there many small `aten::copy_` calls?** Possible CPU↔GPU data transfer bottleneck — check `move_to_device` and where tensors are created.
3. **Does dataloader iteration show up in the CPU timeline?** If the CPU is blocked waiting for the next batch, consider `num_workers > 0` on the `DataLoader` in `make_dataloader`.
4. **Is `loss.item()` a significant fraction of time?** Each `.item()` call synchronizes the GPU. In the current trainer there are two per batch (`loss_value` and inside callbacks) — this is usually fine but worth knowing.

### Minimal one-off profile (no schedule)

If you just want a quick answer without modifying the trainer permanently, the simplest approach is to wrap only a handful of batches using `islice`:

```python
from itertools import islice

with torch.profiler.profile(activities=[...]) as prof:
    for batch_idx, batch in islice(batch_iter, 10):
        optimizer.zero_grad()
        train_step_output = model.training_step(batch)
        self.fabric.backward(train_step_output.loss)
        optimizer.step()

print(prof.key_averages().table(sort_by="cuda_time_total", row_limit=20))
prof.export_chrome_trace("trace.json")
```

This profiles exactly 10 batches and exits, leaving the rest of the training run untouched.

**Comment by JansenNye** [comment]
LGTM (let's give Tostitos to Mo)
